File: 2020/20_JurassicJigsaw/image.test.ts
// ======================================================================
// Jurassic Jigsaw
//   Advent of Code 2020 Day 20 -- Eric Wastl -- https://adventofcode.com
//
// Typescript implementation by Dr. Dean Earl Wright III
//  ======================================================================

// ======================================================================
//                      i m a g e . t e s t . t s
//
// Test the solver for Advent of Code 2020 day 20 problem
// ======================================================================

// ----------------------------------------------------------------------
//                                                                 import
// ----------------------------------------------------------------------

import { fromText } from './aoc_20';
import { Image } from './image';

// ----------------------------------------------------------------------
//                                                              constants
// ----------------------------------------------------------------------
const EXAMPLE_TEXT = `
Tile 2311:
..##.#..#.
##..#.....
#...##..#.
####.#...#
##.##.###.
##...#.###
.#.#.#..##
..#....#..
###...#.#.
..###..###

Tile 1951:
#.##...##.
#.####...#
.....#..##
#...######
.##.#....#
.###.#####
###.##.##.
.###....#.
..#.#..#.#
#...##.#..

Tile 1171:
####...##.
#..##.#..#
##.#..#.#.
.###.####.
..###.####
.##....##.
.#...####.
#.##.####.
####..#...
.....##...

Tile 1427:
###.##.#..
.#..#.##..
.#.##.#..#
#.#.#.##.#
....#...##
...##..##.
...#.#####
.#.####.#.
..#..###.#
..##.#..#.

Tile 1489:
##.#.#....
..##...#..
.##..##...
..#...#...
#####...#.
#..#.#.#.#
...#.#.#..
##.#...##.
..##.##.##
###.##.#..

Tile 2473:
#....####.
#..#.##...
#.##..#...
######.#.#
.#...#.#.#
.#########
.###.#..#.
########.#
##...##.#.
..###.#.#.

Tile 2971:
..#.#....#
#...###...
#.#.###...
##.##..#..
.#####..##
.#..####.#
#..#.#..#.
..####.###
..#.#.###.
...#.#.#.#

Tile 2729:
...#.#.#.#
####.#....
..#.#.....
....#..#.#
.##..##.#.
.#.####...
####.#.#..
##.####...
##..#.##..
#.##...##.

Tile 3079:
#.#.#####.
.#..######
..#.......
######....
####.#..#.
.#...#.##.
#.#####.##
..#.###...
..#.......
..#.###...
`;

const PART_ONE_TEXT = EXAMPLE_TEXT;
const PART_TWO_TEXT = EXAMPLE_TEXT;

const PART_ONE_RESULT = 20899048083289;
const PART_TWO_RESULT = 273;

const EXAMPLE_IMAGE = `.#.#..#.##...#.##..#####
###....#.#....#..#......
##.##.###.#.#..######...
###.#####...#.#####.#..#
##.#....#.##.####...#.##
...########.#....#####.#
....#..#...##..#.#.###..
.####...#..#.....#......
#..#.##..#..###.#.##....
#.####..#.####.#.#.###..
###.#.#...#.######.#..##
#.####....##..########.#
##..##.#...#...#.#.#.#..
...#..#..#.#.##..###.###
.#.#....#.##.#...###.##.
###.#...#..#.##.######..
.#.#.###.##.##.#..#.##..
.####.###.#...###.#..#.#
..#.#..#..#.#.#.####.###
#..####...#.#.#.###.###.
#####..#####...###....##
#.##..#..#...#..####...#
.#.###..##..##..####.##.
...###...##...#...#..###`;

const IMAGE_ROWS = [
  '.#.#..#.##...#.##..#####',
  '###....#.#....#..#......',
  '##.##.###.#.#..######...',
  '###.#####...#.#####.#..#',
  '##.#....#.##.####...#.##',
  '...########.#....#####.#',
  '....#..#...##..#.#.###..',
  '.####...#..#.....#......',
  '#..#.##..#..###.#.##....',
  '#.####..#.####.#.#.###..',
  '###.#.#...#.######.#..##',
  '#.####....##..########.#',
  '##..##.#...#...#.#.#.#..',
  '...#..#..#.#.##..###.###',
  '.#.#....#.##.#...###.##.',
  '###.#...#..#.##.######..',
  '.#.#.###.##.##.#..#.##..',
  '.####.###.#...###.#..#.#',
  '..#.#..#..#.#.#.####.###',
  '#..####...#.#.#.###.###.',
  '#####..#####...###....##',
  '#.##..#..#...#..####...#',
  '.#.###..##..##..####.##.',
  '...###...##...#...#..###',
];

const MONSTER_ROWS = [
  '.####...#####..#...###..',
  '#####..#..#.#.####..#.#.',
  '.#.#...#.###...#.##.##..',
  '#.#.##.###.#.##.##.#####',
  '..##.###.####..#.####.##',
  '...#.#..##.##...#..#..##',
  '#.##.#..#.#..#..##.#.#..',
  '.###.##.....#...###.#...',
  '#.####.#.#....##.#..#.#.',
  '##...#..#....#..#...####',
  '..#.##...###..#.#####..#',
  '....#.##.#.#####....#...',
  '..##.##.###.....#.##..#.',
  '#...#...###..####....##.',
  '.#.##...#.##.#.#.###...#',
  '#.###.#..####...##..#...',
  '#.###...#.##...#.######.',
  '.###.###.#######..#####.',
  '..##.#..#..#.#######.###',
  '#.#..##.########..#..##.',
  '#.#####..#.#...##..#....',
  '#....##..#.#########..##',
  '#...#.....#..##...###.##',
  '#..###....##.#...##.##.#',
];

// ======================================================================
//                                                              TestImage
// ======================================================================

describe('Image', () => {
  test('Test the default Image creation', () => {
    // 1. Create default Image object
    const myobj = new Image([]);
    // 2. Make sure it has the default values
    expect(myobj.part2).toBe(false);
    expect(myobj.text).toHaveLength(0);
    expect(myobj.tiles.tiles).toHaveLength(0);
    expect(myobj.image).toHaveLength(0);
  });

  test('Test the Image object creation from text', () => {
    // 1. Create Image object from text
    const myobj = new Image(fromText(EXAMPLE_TEXT));
    // 2. Make sure it has the expected values
    expect(myobj.part2).toBe(false);
    expect(myobj.text).toHaveLength(11 * 9);
    expect(myobj.tiles.tiles).toHaveLength(9);
    expect(myobj.image).toHaveLength(0);

    // 3. Check methods
    expect(myobj.corners()).toBe(PART_ONE_RESULT);
    expect(myobj.countPoundSigns(EXAMPLE_IMAGE.split('\n'))).toBe(303);
    expect(myobj.countPoundSigns(IMAGE_ROWS)).toBe(303);
    expect(myobj.countPoundSigns(MONSTER_ROWS)).toBe(303);
    expect(myobj.countMonsters(MONSTER_ROWS)).toBe(2);
  });

  test('Test part one example of Image object', () => {
    // 1. Create Image object from text
    const myobj = new Image(fromText(PART_ONE_TEXT));
    // 2. Check the part one result
    expect(myobj.partOne(false, 0)).toBe(PART_ONE_RESULT);
  });

  test('Test part two example of Image object', () => {
    // 1. Create Image object from text
    const myobj = new Image(fromText(PART_TWO_TEXT), true);
    // 2. Check the part two result
    expect(myobj.partTwo(false, 0)).toBe(PART_TWO_RESULT);
  });
});

// ======================================================================
// end                    i m a g e . t e s t . t s                   end
// ======================================================================

File: 2023/11_CosmicExpansion/image_test.go
// ======================================================================
// Cosmic Expansion
//   Advent of Code 2023 Day 11 -- Eric Wastl -- https://adventofcode.com
//
// go implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                    i m a g e _ t e s t . g o
// ======================================================================
// Test solver for Advent of Code 2023 day 11, Cosmic Expansion

package main

// ----------------------------------------------------------------------
//                                                                imports
// ----------------------------------------------------------------------
import (
	"testing"

	"github.com/stretchr/testify/assert"
)

// ----------------------------------------------------------------------
//
//	constants
//
// ----------------------------------------------------------------------
const (
	imageExampleText = `...#......
.......#..
#.........
..........
......#...
.#........
.........#
..........
.......#..
#...#.....
`
	imagePartOneText = imageExampleText
	imagePartTwoText = ""

	imagePartOneResult = "374"
	imagePartTwoResult = ""
)

// ======================================================================
//                                                              TestImage
// ======================================================================

// TestImageEmptyInit ... Test the default Image creation
func TestImageEmptyInit(t *testing.T) {

	// 1. Create default Image object
	obj, err := NewImage(false, fromText(""))
	assert.NoError(t, err)

	// 2. Make sure it has the default values
	assert.False(t, obj.Part2)
	assert.Len(t, obj.Text, 0)
	assert.Len(t, obj.Galaxies, 0)
}

// TestImageTextInit ... Test the Image object creation from text
func TestImageTextInit(t *testing.T) {

	// 1. Create Image object from text
	obj, err := NewImage(false, fromText(imageExampleText))
	assert.NoError(t, err)

	// 2. Make sure it has the expected values
	assert.False(t, obj.Part2)
	assert.Len(t, obj.Text, 10)
	assert.Len(t, obj.Galaxies, 9)

	// 3. Check methods
	assert.Equal(t, int64(5), obj.Galaxies[7].DistanceAll(obj.Galaxies[8:]))
	assert.Equal(t, int64(15), obj.Galaxies[6].DistanceAll(obj.Galaxies[7:]))
	assert.Equal(t, int64(33), obj.Galaxies[5].DistanceAll(obj.Galaxies[6:]))
	assert.Equal(t, int64(39), obj.Galaxies[4].DistanceAll(obj.Galaxies[5:]))
	assert.Equal(t, int64(43), obj.Galaxies[3].DistanceAll(obj.Galaxies[4:]))
	assert.Equal(t, int64(73), obj.Galaxies[2].DistanceAll(obj.Galaxies[3:]))
	assert.Equal(t, int64(79), obj.Galaxies[1].DistanceAll(obj.Galaxies[2:]))
	assert.Equal(t, int64(87), obj.Galaxies[0].DistanceAll(obj.Galaxies[1:]))
	assert.Equal(t, int64(374), obj.DistanceAll())

}

// TestImagePartOne ... Test part one example of Image object
func TestImagePartOne(t *testing.T) {

	// 1. Create Image object from text
	obj, err := NewImage(false, fromText(imagePartOneText))
	assert.NoError(t, err)

	// 2. Make sure it has the expected values
	assert.False(t, obj.Part2)
	assert.Len(t, obj.Text, 10)
	assert.Len(t, obj.Galaxies, 9)

	// 3. Make sure it returns the expected solution
	result := obj.PartOne(false, 0)
	assert.Equal(t, imagePartOneResult, result)
}

// TestImagePartTwo ... Test part two example of Image object
func TestImagePartTwo(t *testing.T) {
	// Test part two example of Image object

	// 1. Create Image object from text
	obj, err := NewImage(true, fromText(imagePartTwoText))
	assert.NoError(t, err)

	// 2. Make sure it has the expected values
	assert.True(t, obj.Part2)
	assert.Len(t, obj.Text, 10)
	assert.Len(t, obj.Galaxies, 9)

	// 3. Make sure it returns the expected solution
	result := obj.PartTwo(false, 0)
	assert.Equal(t, imagePartTwoResult, result)
}

// ======================================================================
// end                   i m a g e _ t e s t . g o                    end
// ======================================================================
